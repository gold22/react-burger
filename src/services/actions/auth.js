export const REGISTER_USER_REQUEST = 'REGISTER_USER_REQUEST';
export const REGISTER_USER_SUCCESS = 'REGISTER_USER_SUCCESS';
export const REGISTER_USER_ERROR = 'REGISTER_USER_ERROR';

export const registerUser = (user) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result.user,
            accessToken: result.accessToken,
            refreshToken: result.refreshToken,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: error.message,
        });
    }
};
