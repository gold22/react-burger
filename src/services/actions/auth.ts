import {
    REGISTER_USER_REQUEST,
    REGISTER_USER_SUCCESS,
    REGISTER_USER_ERROR,
    LOGIN_USER_REQUEST,
    LOGIN_USER_SUCCESS,
    LOGIN_USER_ERROR,
    LOGOUT_USER_REQUEST,
    LOGOUT_USER_SUCCESS,
    LOGOUT_USER_ERROR,
    GET_USER_REQUEST,
    GET_USER_SUCCESS,
    GET_USER_ERROR,
    UPDATE_USER_REQUEST,
    UPDATE_USER_SUCCESS,
    UPDATE_USER_ERROR,
} from '../constants/auth';
import { TApiUser, TApiUserCredentials } from '../types/api';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

export const registerUser: TThunk = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logInUser: TThunk = (user: TApiUserCredentials) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGIN_USER_REQUEST,
    });
    try {
        const result = await apiClient.logInUser(user);
        dispatch({
            type: LOGIN_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: LOGIN_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logOutUser: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGOUT_USER_REQUEST,
    });
    try {
        await apiClient.logOutUser();
        dispatch({
            type: LOGOUT_USER_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: LOGOUT_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const getUser = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_USER_REQUEST,
    });
    try {
        const user = await apiClient.getUser();
        dispatch({
            type: GET_USER_SUCCESS,
            user,
        });
    } catch (error) {
        dispatch({
            type: GET_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const updateUser = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: UPDATE_USER_REQUEST,
    });
    try {
        const result = await apiClient.updateUser(user);
        dispatch({
            type: UPDATE_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: UPDATE_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
