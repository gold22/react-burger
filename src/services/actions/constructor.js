import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';

export const setBun = (bun) => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient, index) => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index) => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = () => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex, newIndex) => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
