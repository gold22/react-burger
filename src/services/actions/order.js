import { getIngredientsIds } from '../../utils/ingredients';

export const CREATE_ORDER_REQUEST = 'CREATE_ORDER_REQUEST';
export const CREATE_ORDER_SUCCESS = 'CREATE_ORDER_SUCCESS';
export const CREATE_ORDER_ERROR = 'CREATE_ORDER_ERROR';

export const createOrder = () => async (dispatch, getState, apiClient) => {
    dispatch({
        type: CREATE_ORDER_REQUEST,
    });
    try {
        const { ingredients } = getState().burgerConstructor;
        const result = await apiClient.createOrder(getIngredientsIds(ingredients));
        dispatch({
            type: CREATE_ORDER_SUCCESS,
            name: result.name,
            number: result.order.number,
        });
    } catch (error) {
        dispatch({
            type: CREATE_ORDER_ERROR,
            message: error.message,
        });
    }
};
