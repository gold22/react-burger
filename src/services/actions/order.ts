import { getIngredientsIds } from '../../utils/ingredients';
import {
    CREATE_ORDER_REQUEST,
    CREATE_ORDER_SUCCESS,
    CREATE_ORDER_ERROR,
} from '../constants/order';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

// eslint-disable-next-line import/prefer-default-export
export const createOrder: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: CREATE_ORDER_REQUEST,
    });
    try {
        const { ingredients } = getState().burgerConstructor;
        const result = await apiClient.createOrder(getIngredientsIds(ingredients));
        dispatch({
            type: CREATE_ORDER_SUCCESS,
            name: result.name,
            number: result.order.number,
        });
    } catch (error) {
        dispatch({
            type: CREATE_ORDER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
