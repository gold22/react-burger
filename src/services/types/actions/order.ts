import {
    CREATE_ORDER_REQUEST,
    CREATE_ORDER_SUCCESS,
    CREATE_ORDER_ERROR,
} from '../../constants/order';

export interface ICreateOrderAction {
    readonly type: typeof CREATE_ORDER_REQUEST;
}

export interface ICreateOrderSuccessAction {
    readonly type: typeof CREATE_ORDER_SUCCESS;
    readonly name: string;
    readonly number: string;
}

export interface ICreateOrderFailedAction {
    readonly type: typeof CREATE_ORDER_ERROR;
    readonly message: string;
}

export type TOrderActions =
    | ICreateOrderAction
    | ICreateOrderSuccessAction
    | ICreateOrderFailedAction;
