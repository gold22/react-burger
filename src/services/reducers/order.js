import * as actions from '../actions/order';

const initialState = {
    name: null,
    number: null,
    isCreating: false,
    creationError: null,
};

// eslint-disable-next-line import/prefer-default-export
export const orderReducer = (state = initialState, action) => {
    switch (action.type) {
    case actions.CREATE_ORDER_REQUEST: {
        return {
            ...state,
            isCreating: true,
            creationError: null,
        };
    }
    case actions.CREATE_ORDER_SUCCESS: {
        return {
            ...state,
            name: action.name,
            number: action.number,
            isCreating: false,
        };
    }
    case actions.CREATE_ORDER_ERROR: {
        return {
            ...state,
            isCreating: false,
            creationError: action.message,
        };
    }
    default: {
        return state;
    }
    }
};
