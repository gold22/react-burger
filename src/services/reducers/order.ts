import {
    CREATE_ORDER_REQUEST,
    CREATE_ORDER_SUCCESS,
    CREATE_ORDER_ERROR,
} from '../constants/order';
import { TOrderActions } from '../types/actions/order';

export type TOrderState = {
    name: string | null;
    number: string | null;
    isCreating: boolean;
    creationError: string | null;
};

const initialState: TOrderState = {
    name: null,
    number: null,
    isCreating: false,
    creationError: null,
};

export const orderReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TOrderState = initialState,
    action: TOrderActions,
) => {
    switch (action.type) {
    case CREATE_ORDER_REQUEST: {
        return {
            ...state,
            isCreating: true,
        };
    }
    case CREATE_ORDER_SUCCESS: {
        return {
            ...state,
            name: action.name,
            number: action.number,
            isCreating: false,
            creationError: null,
        };
    }
    case CREATE_ORDER_ERROR: {
        return {
            ...initialState,
            creationError: action.message,
        };
    }
    default: {
        return state;
    }
    }
};
