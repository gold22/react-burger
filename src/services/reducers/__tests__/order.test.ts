import { orderReducer, TOrderState } from '../order';
import {
    CREATE_ORDER_ERROR,
    CREATE_ORDER_REQUEST,
    CREATE_ORDER_SUCCESS,
} from '../../constants/order';
import { ICreateOrderSuccessAction } from '../../types/actions/order';

const initialState: TOrderState = orderReducer(undefined, {} as any);

describe('order reducer', () => {
    it('should return the initial state', () => {
        expect(initialState).toMatchSnapshot();
    });

    it('should handle CREATE_ORDER_REQUEST', () => {
        expect(orderReducer(undefined, { type: CREATE_ORDER_REQUEST }))
            .toMatchSnapshot();
    });

    it('should handle CREATE_ORDER_SUCCESS', () => {
        const action: ICreateOrderSuccessAction = {
            type: CREATE_ORDER_SUCCESS,
            name: 'Традиционный-галактический краторный бургер',
            number: '9932',
        };
        const state = {
            ...initialState,
            isCreating: true,
        };
        expect(orderReducer(state, action))
            .toMatchSnapshot();
    });

    it('should handle CREATE_ORDER_ERROR', () => {
        const state = {
            ...initialState,
            isCreating: true,
        };
        expect(orderReducer(state, { type: CREATE_ORDER_ERROR, message: 'Create order error' }))
            .toMatchSnapshot();
    });
});
