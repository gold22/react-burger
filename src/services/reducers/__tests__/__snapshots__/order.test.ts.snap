// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order reducer should handle CREATE_ORDER_ERROR 1`] = `
Object {
  "creationError": "Create order error",
  "isCreating": false,
  "name": null,
  "number": null,
}
`;

exports[`order reducer should handle CREATE_ORDER_REQUEST 1`] = `
Object {
  "creationError": null,
  "isCreating": true,
  "name": null,
  "number": null,
}
`;

exports[`order reducer should handle CREATE_ORDER_SUCCESS 1`] = `
Object {
  "creationError": null,
  "isCreating": false,
  "name": "Традиционный-галактический краторный бургер",
  "number": "9932",
}
`;

exports[`order reducer should return the initial state 1`] = `
Object {
  "creationError": null,
  "isCreating": false,
  "name": null,
  "number": null,
}
`;
